src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -3173,11 +3173,15 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
   CompiledEntrySignature ces(method());
   ces.compute_calling_conventions();
   if (ces.has_scalarized_args()) {
-    method->set_has_scalarized_args();
+    if (!method->has_scalarized_args()) {
+      assert(!method()->constMethod()->is_shared(), "Cannot update shared const object");
```

Comment:
Is this assertion mandatory ? since shared archives are mapped with MAP_PRIVATE flag [ (on linux)](https://github.com/openjdk/valhalla/blob/lworld/src/hotspot/os/linux/os_linux.cpp#L3418) thus it  follow COW semantics, any update to shared mapped region will clone the physical frame before update.

Reply:
Shared archives are split in two regions. The writable part is using the COW semantic, but all the "const*" metadata are in an unique read-only region.